### PR TITLE
Exports + .d.ts emit

### PR DESCRIPTION
**Pull request by vladima**
This PR is an aggregation of #2008 and #2025 and subsumes them.

### CODE REVIEW

**Comment by csnover** [comment]
Sorry to bother, do you know when this will be landing on master? It’s necessary for me to update dts-generator for use with ES6 modules.

**Comment by mhegazy** [comment]
Hopefully by end of week. Sheetal has been sick for a while now, @vladima is picking it up.

**Comment by csnover** [comment]
Great, thanks!

I hope you feel better soon, @sheetalkamat ! Being sick is no fun. :mask:

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1837,6 +1861,40 @@ module ts {
             }
         }
 
+        function setDeclarationsOfIdentifierAsVisible(node: Identifier): Node[]{
```
Comment (mhegazy) [inline src/compiler/checker.ts]: can we find a better name.

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -1649,6 +1649,10 @@
         "category": "Error",
         "code": 4081
     },
+    "Default export of the module has or is using private name '{0}'.": {
+        "category": "Error",
+        "code": 4081
```
Comment (mhegazy) [inline src/compiler/diagnosticMessages.json]: update code

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -40,16 +40,18 @@ module ts {
...
 
-    interface AliasDeclarationEmitInfo {
-        declaration: ImportEqualsDeclaration;
+    interface ModuleElementDeclarationEmitInfo {
+        node: Node;
         outputPos: number;
         indent: number;
         asynchronousOutput?: string; // If the output for alias was written asynchronously, the corresponding output
+        subModuleElementDeclarationEmitInfo?: ModuleElementDeclarationEmitInfo[];
+        isVisible?: boolean;
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: may be call this isReferenced or isUsed to distinguish it from the checker's isVisible flag..

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -374,7 +376,8 @@ module ts {
         let emitJsDocComments = compilerOptions.removeComments ? function (declaration: Node) { } : writeJsDocComments;
         let emit = compilerOptions.stripInternal ? stripInternal : emitNode;
 
-        let aliasDeclarationEmitInfo: AliasDeclarationEmitInfo[] = [];
+        let moduleElementDeclarationEmitInfo: ModuleElementDeclarationEmitInfo[] = [];
+        let asynchronousSubModuleDeclarationEmitInfo: ModuleElementDeclarationEmitInfo[];
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: also needs a new name.. possibly pendingEdits

File: tests/cases/compiler/es6ImportDefaultBindingFollowedWithNamedImportDts.ts.orig [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<<<<<<< HEAD
+// @target: es6
```
Comment (mhegazy) [inline tests/cases/compiler/es6ImportDefaultBindingFollowedWithNamedImportDts.ts.orig]: should be droped

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**